### PR TITLE
#287 Dhaden/flush return rejectable promise

### PR DESCRIPTION
**Pull request by Dahaden**
Referencing https://github.com/segmentio/analytics-node/issues/287

- Return rejected promises via flush
- Attempted to fix tests (not quite there)

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -1,12 +1,12 @@
-import { spy, stub } from 'sinon'
-import bodyParser from 'body-parser'
-import express from 'express'
-import delay from 'delay'
-import auth from 'basic-auth'
-import pify from 'pify'
-import test from 'ava'
-import Analytics from '.'
-import { version } from './package'
+const { spy, stub } = require('sinon')
```
Comment (Dahaden) [inline test.js:1]: Annoyingly bable was removed from ava and imports arent native until node 13 so have to revert to requires here

File: test.js [hunk 2]
```diff
@@ -88,7 +88,8 @@ test('expose a constructor', t => {
 })
 
 test('require a write key', t => {
-  t.throws(() => new Analytics(), 'You must pass your Segment project\'s write key.')
+  const error = t.throws(() => new Analytics(), { code: 'ERR_ASSERTION' })
+  t.is(error.message, 'You must pass your Segment project\'s write key.')
```
Comment (Dahaden) [inline test.js:92]: With latest versions of ava, you can no longer provide just a string.

I can make this consistent with the other tests, or use this method for testing exceptions. Either or.

File: test.js [hunk 3]
```diff
@@ -296,7 +297,7 @@ test('enqueue - skip when client is disabled', async t => {
 test('flush - don\'t fail when queue is empty', async t => {
   const client = createClient()
 
-  await t.notThrows(client.flush())
+  await t.notThrowsAsync(() => client.flush())
```
Comment (Dahaden) [inline test.js:300]: For anything that returns a promise, we have to use `t.throwsAsync` or `t.notThrowsAsync` now

File: test.js [hunk 4]
```diff
@@ -342,7 +343,11 @@ test('flush - respond with an error', async t => {
     }
   ]
 
-  await t.throws(client.flush(), 'Bad Request')
+  try {
+    await t.throwsAsync(() => client.flush())
+  } catch (error) {
+    console.log('Dafuq:', error)
```
Comment (Dahaden) [inline test.js:349]: Part of debugging in an attempt to find where the unhandled promise rejection was coming from.
This console log is not called at all. Very confused

**Comment by Dahaden** [comment]
Closing this PR in favour of https://github.com/segmentio/analytics-node/pull/294

**Comment by pbassut** [comment]
@Dahaden But this PR included some useful upgrades(like ava and babel), right? 
I don't think we should ditch that

**Comment by Dahaden** [comment]
I can create another PR with just the ava bump if you want? The big changes are just:

1. No import (because babel got removed from ava, might be a better way to do that though?)
2. Change in API for Async functions,
3. Change in API for error matching

**Comment by pbassut** [comment]
That would be terrific @Dahaden . 
We're on a really old version of Ava.

**Comment by pbassut** [comment]
I like the baby steps approach here. Let's get this #288 going as it is the biggest chunk of the whole deal.

**Pull request closed without merging**